File: packages/usfm-parser-integration-tests/tests/bsb.test.ts
import { describe, it } from "@std/testing/bdd";
import { expect } from "@std/expect";
import { parse } from "@usfm-tools/parser";
import type {
  BookNode,
  ChapterNode,
  VerseNode,
  ParagraphNode,
  CharNode,
  NoteNode,
  TextNode,
  UsfmNode,
  ParentNode,
  ParseResult,
} from "@usfm-tools/parser";
const BSB_DIR = new URL("../../../bibles/bsb/usfm/", import.meta.url);

function loadUsfm(filename: string): string {
  return Deno.readTextFileSync(new URL(filename, BSB_DIR));
}

function allUsfmFiles(): string[] {
  const files: string[] = [];
  for (const entry of Deno.readDirSync(BSB_DIR)) {
    if (entry.isFile && entry.name.endsWith(".usfm")) {
      files.push(entry.name);
    }
  }
  return files.sort();
}

function findChild<T>(
  children: { type: string }[],
  type: string,
  predicate?: (node: T) => boolean,
): T | undefined {
  return children.find(
    (c) => c.type === type && (!predicate || predicate(c as T)),
  ) as T | undefined;
}

function filterChildren<T>(children: { type: string }[], type: string): T[] {
  return children.filter((c) => c.type === type) as T[];
}

function collectText(children: { type: string; text?: string }[]): string {
  return children
    .filter((c) => c.type === "text")
    .map((c) => (c as TextNode).text)
    .join("");
}

/**
 * Collect sibling nodes that follow a verse milestone in a parent container.
 * Walks all paragraphs in the parent looking for the verse, then returns
 * siblings until the next verse or end of paragraph.
 */
function collectAfterVerse(parent: ParentNode, verseNumber: string): UsfmNode[] {
  for (const child of parent.children) {
    if (!("children" in child)) continue;
    const paraChildren = (child as ParentNode).children;
    let collecting = false;
    const result: UsfmNode[] = [];
    for (const node of paraChildren) {
      if (node.type === "verse") {
        if ((node as VerseNode).number === verseNumber) {
          collecting = true;
          continue;
        } else if (collecting) {
          return result;
        }
      }
      if (collecting) result.push(node);
    }
    if (collecting) return result;
  }
  return [];
}

// ---------------------------------------------------------------------------
// Test suite
// ---------------------------------------------------------------------------

describe("BSB Integration Tests", () => {
  const files = allUsfmFiles();

  // -------------------------------------------------------------------------
  // 1. All 66 books parse without throwing
  // -------------------------------------------------------------------------

  describe("parse all books without exceptions", () => {
    it("should find exactly 66 USFM files", () => {
      expect(files).toHaveLength(66);
    });

    it("should parse every book in non-strict mode without throwing", () => {
      for (const file of files) {
        const content = loadUsfm(file);
        expect(() => parse(content)).not.toThrow();
      }
    });
  });

  // -------------------------------------------------------------------------
  // 2. Performance: entire Bible under 30 seconds
  // -------------------------------------------------------------------------

  describe("performance", () => {
    it("should parse the entire BSB in under 30 seconds", () => {
      const start = performance.now();
      for (const file of files) {
        const content = loadUsfm(file);
        parse(content);
      }
      const elapsed = performance.now() - start;
      expect(elapsed).toBeLessThan(30_000);
    });
  });

  // -------------------------------------------------------------------------
  // 3. Sampled structural assertions
  // -------------------------------------------------------------------------

  describe("sampled output: Genesis", () => {
    let result: ParseResult;

    it("should parse without throwing", () => {
      result = parse(loadUsfm("GEN.usfm"));
    });

    it("should have a book node with code GEN", () => {
      const book = result.document.children[0] as BookNode;
      expect(book.type).toBe("book");
      expect(book.code).toBe("GEN");
    });

    it("should have 50 chapters", () => {
      const book = result.document.children[0] as BookNode;
      const chapters = filterChildren<ChapterNode>(book.children, "chapter");
      expect(chapters).toHaveLength(50);
      expect(chapters[0].number).toBe("1");
      expect(chapters[49].number).toBe("50");
    });

    it("should have verse 1 in chapter 1 with expected text", () => {
      const book = result.document.children[0] as BookNode;
      const ch1 = findChild<ChapterNode>(book.children, "chapter");
      expect(ch1).toBeDefined();

      // Verse is a milestone — collect text siblings after it
      const afterV1 = collectAfterVerse(ch1!, "1");
      expect(afterV1.length).toBeGreaterThan(0);
      const text = collectText(afterV1);
      expect(text).toContain("In the beginning God created");
    });

    it("should contain a section heading 'The Creation'", () => {
      const book = result.document.children[0] as BookNode;
      const ch1 = findChild<ChapterNode>(book.children, "chapter");

      const s1 = findChild<ParagraphNode>(
        ch1!.children,
        "paragraph",
        (p: ParagraphNode) => p.marker === "s1",
      );
      expect(s1).toBeDefined();
      const text = collectText(s1!.children);
      expect(text).toContain("The Creation");
    });
  });

  describe("sampled output: Psalms", () => {
    let result: ParseResult;

    it("should parse without throwing", () => {
      result = parse(loadUsfm("PSA.usfm"));
    });

    it("should have book code PSA with 150 chapters", () => {
      const book = result.document.children[0] as BookNode;
      expect(book.code).toBe("PSA");
      const chapters = filterChildren<ChapterNode>(book.children, "chapter");
      expect(chapters).toHaveLength(150);
    });

    it("should have poetry formatting (q1/q2 markers)", () => {
      const book = result.document.children[0] as BookNode;
      const ch1 = findChild<ChapterNode>(book.children, "chapter");
      const paras = filterChildren<ParagraphNode>(ch1!.children, "paragraph");
      const poetryMarkers = paras
        .map((p) => p.marker)
        .filter((m) => m.startsWith("q"));
      expect(poetryMarkers.length).toBeGreaterThan(0);
      expect(poetryMarkers).toContain("q1");
      expect(poetryMarkers).toContain("q2");
    });

    it("should have 'Blessed is the man' in Psalm 1:1", () => {
      const book = result.document.children[0] as BookNode;
      const ch1 = findChild<ChapterNode>(book.children, "chapter");
      expect(ch1).toBeDefined();

      const afterV1 = collectAfterVerse(ch1!, "1");
      expect(afterV1.length).toBeGreaterThan(0);
      const text = collectText(afterV1);
      expect(text).toContain("Blessed is the man");
    });
  });

  describe("sampled output: Matthew", () => {
    let result: ParseResult;

    it("should parse without throwing", () => {
      result = parse(loadUsfm("MAT.usfm"));
    });

    it("should have book code MAT with 28 chapters", () => {
      const book = result.document.children[0] as BookNode;
      expect(book.code).toBe("MAT");
      const chapters = filterChildren<ChapterNode>(book.children, "chapter");
      expect(chapters).toHaveLength(28);
    });

    it("should have a header 'Matthew'", () => {
      const book = result.document.children[0] as BookNode;
      const header = findChild<ParagraphNode>(
        book.children,
        "paragraph",
        (p: ParagraphNode) => p.marker === "h",
      );
      expect(header).toBeDefined();
      const text = collectText(header!.children);
      expect(text).toContain("Matthew");
    });

    it("should contain footnotes in chapter 1", () => {
      const book = result.document.children[0] as BookNode;
      const ch1 = findChild<ChapterNode>(book.children, "chapter");

      const notes: NoteNode[] = [];
      const walk = (children: { type: string }[]) => {
        for (const c of children) {
          if (c.type === "note") notes.push(c as NoteNode);
          if ("children" in c) walk((c as { children: { type: string }[] }).children);
        }
      };
      walk(ch1!.children);

      expect(notes.length).toBeGreaterThan(0);
      expect(notes[0].marker).toBe("f");
      expect(notes[0].caller).toBe("+");
    });
  });

  describe("sampled output: Revelation (last book)", () => {
    let result: ParseResult;

    it("should parse without throwing", () => {
      result = parse(loadUsfm("REV.usfm"));
    });

    it("should have book code REV with 22 chapters", () => {
      const book = result.document.children[0] as BookNode;
      expect(book.code).toBe("REV");
      const chapters = filterChildren<ChapterNode>(book.children, "chapter");
      expect(chapters).toHaveLength(22);
    });

    it("should contain \\wj (words of Jesus) char nodes", () => {
      const book = result.document.children[0] as BookNode;
      const ch1 = findChild<ChapterNode>(book.children, "chapter");

      const wjNodes: CharNode[] = [];
      const walk = (children: { type: string }[]) => {
        for (const c of children) {
          if (c.type === "char" && (c as CharNode).marker === "wj") {
            wjNodes.push(c as CharNode);
          }
          if ("children" in c) walk((c as { children: { type: string }[] }).children);
        }
      };
      walk(ch1!.children);

      expect(wjNodes.length).toBeGreaterThan(0);
    });
  });

  // -------------------------------------------------------------------------
  // 4. Error analysis — characterize parser limitations vs real USFM errors
  // -------------------------------------------------------------------------

  describe("error characterization", () => {
    it("should parse all 66 BSB books with zero errors", () => {
      let totalErrors = 0;
      const sampleErrors: { file: string; message: string }[] = [];

      for (const file of files) {
        const content = loadUsfm(file);
        const result = parse(content);
        totalErrors += result.errors.length;
        for (const e of result.errors.slice(0, 3)) {
          sampleErrors.push({ file, message: e.message });
        }
      }

      if (sampleErrors.length > 0) {
        console.log("Unexpected errors:", sampleErrors);
      }
      expect(totalErrors).toBe(0);
    });
  });
});

File: usfm-parser-go/integration/bsb_test.go
// Package integration runs the Berean Standard Bible corpus tests against
// the Go parser. Port of packages/usfm-parser-integration-tests/tests/bsb.test.ts.
package integration

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
	"testing"
	"time"

	usfm "github.com/usfm-tools/usfm-parser-go"
	"github.com/usfm-tools/usfm-parser-go/parser"
)

const bsbDir = "../../bibles/bsb/usfm"

func allUsfmFiles(t *testing.T) []string {
	t.Helper()
	entries, err := os.ReadDir(bsbDir)
	if err != nil {
		t.Skipf("BSB corpus not available: %v", err)
	}
	var files []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".usfm") {
			files = append(files, e.Name())
		}
	}
	sort.Strings(files)
	return files
}

func loadUsfm(t *testing.T, filename string) string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(bsbDir, filename))
	if err != nil {
		t.Fatalf("read %s: %v", filename, err)
	}
	return string(data)
}

func parseBook(t *testing.T, filename string) usfm.ParseResult {
	t.Helper()
	result := parser.Parse(loadUsfm(t, filename))
	if len(result.Document.Children) == 0 {
		t.Fatalf("%s: document has no children", filename)
	}
	return result
}

func findChild(parent *usfm.Node, typ usfm.NodeType) *usfm.Node {
	for _, c := range parent.Children {
		if c.Type == typ {
			return c
		}
	}
	return nil
}

func filterChildren(parent *usfm.Node, typ usfm.NodeType) []*usfm.Node {
	var out []*usfm.Node
	for _, c := range parent.Children {
		if c.Type == typ {
			out = append(out, c)
		}
	}
	return out
}

func collectText(nodes []*usfm.Node) string {
	var b strings.Builder
	for _, c := range nodes {
		if c.Type == usfm.NodeText {
			b.WriteString(c.Text)
		}
	}
	return b.String()
}

// collectAfterVerse collects sibling nodes that follow a verse milestone:
// it walks the parent's child containers looking for the verse, returning
// siblings until the next verse or end of that container.
func collectAfterVerse(parent *usfm.Node, verseNumber string) []*usfm.Node {
	for _, child := range parent.Children {
		if child.Children == nil {
			continue
		}
		collecting := false
		var result []*usfm.Node
		for _, node := range child.Children {
			if node.Type == usfm.NodeVerse {
				if node.Number == verseNumber {
					collecting = true
					continue
				} else if collecting {
					return result
				}
			}
			if collecting {
				result = append(result, node)
			}
		}
		if collecting {
			return result
		}
	}
	return nil
}

func walk(nodes []*usfm.Node, visit func(*usfm.Node)) {
	for _, n := range nodes {
		visit(n)
		walk(n.Children, visit)
	}
}

// --- 1. All 66 books parse ---

func TestAllBooksPresent(t *testing.T) {
	files := allUsfmFiles(t)
	if len(files) != 66 {
		t.Errorf("found %d USFM files, want 66", len(files))
	}
}

func TestParseAllBooks(t *testing.T) {
	// A panic anywhere in Parse fails the test — the Go analogue of the TS
	// "does not throw in non-strict mode" assertion
	for _, file := range allUsfmFiles(t) {
		parser.Parse(loadUsfm(t, file))
	}
}

// --- 2. Performance ---

func TestParseEntireBibleUnder30Seconds(t *testing.T) {
	files := allUsfmFiles(t)
	contents := make([]string, len(files))
	for i, file := range files {
		contents[i] = loadUsfm(t, file)
	}
	start := time.Now()
	for _, content := range contents {
		parser.Parse(content)
	}
	if elapsed := time.Since(start); elapsed > 30*time.Second {
		t.Errorf("parsing the BSB took %v, want under 30s", elapsed)
	}
}

// --- 3. Sampled structural assertions ---

func TestGenesis(t *testing.T) {
	allUsfmFiles(t) // skip when corpus is missing
	result := parseBook(t, "GEN.usfm")
	book := result.Document.Children[0]
	if book.Type != usfm.NodeBook || book.Code != "GEN" {
		t.Fatalf("book = %v %q, want book GEN", book.Type, book.Code)
	}

	chapters := filterChildren(book, usfm.NodeChapter)
	if len(chapters) != 50 {
		t.Fatalf("chapters = %d, want 50", len(chapters))
	}
	if chapters[0].Number != "1" || chapters[49].Number != "50" {
		t.Errorf("chapter numbers = %q..%q, want 1..50", chapters[0].Number, chapters[49].Number)
	}

	ch1 := chapters[0]
	afterV1 := collectAfterVerse(ch1, "1")
	if len(afterV1) == 0 {
		t.Fatal("no nodes after verse 1")
	}
	if text := collectText(afterV1); !strings.Contains(text, "In the beginning God created") {
		t.Errorf("Gen 1:1 text = %q", text)
	}

	var s1 *usfm.Node
	for _, p := range filterChildren(ch1, usfm.NodeParagraph) {
		if p.Marker == "s1" {
			s1 = p
			break
		}
	}
	if s1 == nil {
		t.Fatal("no s1 section heading in chapter 1")
	}
	if text := collectText(s1.Children); !strings.Contains(text, "The Creation") {
		t.Errorf("s1 text = %q, want to contain \"The Creation\"", text)
	}
}

func TestPsalms(t *testing.T) {
	allUsfmFiles(t)
	result := parseBook(t, "PSA.usfm")
	book := result.Document.Children[0]
	if book.Code != "PSA" {
		t.Fatalf("code = %q, want PSA", book.Code)
	}
	chapters := filterChildren(book, usfm.NodeChapter)
	if len(chapters) != 150 {
		t.Fatalf("chapters = %d, want 150", len(chapters))
	}

	ch1 := chapters[0]
	markers := map[string]bool{}
	for _, p := range filterChildren(ch1, usfm.NodeParagraph) {
		if strings.HasPrefix(p.Marker, "q") {
			markers[p.Marker] = true
		}
	}
	if len(markers) == 0 || !markers["q1"] || !markers["q2"] {
		t.Errorf("poetry markers = %v, want q1 and q2", markers)
	}

	afterV1 := collectAfterVerse(ch1, "1")
	if len(afterV1) == 0 {
		t.Fatal("no nodes after verse 1")
	}
	if text := collectText(afterV1); !strings.Contains(text, "Blessed is the man") {
		t.Errorf("Psalm 1:1 text = %q", text)
	}
}

func TestMatthew(t *testing.T) {
	allUsfmFiles(t)
	result := parseBook(t, "MAT.usfm")
	book := result.Document.Children[0]
	if book.Code != "MAT" {
		t.Fatalf("code = %q, want MAT", book.Code)
	}
	if chapters := filterChildren(book, usfm.NodeChapter); len(chapters) != 28 {
		t.Fatalf("chapters = %d, want 28", len(chapters))
	}

	var header *usfm.Node
	for _, p := range filterChildren(book, usfm.NodeParagraph) {
		if p.Marker == "h" {
			header = p
			break
		}
	}
	if header == nil {
		t.Fatal("no \\h header paragraph")
	}
	if text := collectText(header.Children); !strings.Contains(text, "Matthew") {
		t.Errorf("header text = %q, want to contain Matthew", text)
	}

	ch1 := findChild(book, usfm.NodeChapter)
	var notes []*usfm.Node
	walk(ch1.Children, func(n *usfm.Node) {
		if n.Type == usfm.NodeNote {
			notes = append(notes, n)
		}
	})
	if len(notes) == 0 {
		t.Fatal("no footnotes in Matthew 1")
	}
	if notes[0].Marker != "f" || notes[0].Caller != "+" {
		t.Errorf("first note = marker %q caller %q, want f +", notes[0].Marker, notes[0].Caller)
	}
}

func TestRevelation(t *testing.T) {
	allUsfmFiles(t)
	result := parseBook(t, "REV.usfm")
	book := result.Document.Children[0]
	if book.Code != "REV" {
		t.Fatalf("code = %q, want REV", book.Code)
	}
	if chapters := filterChildren(book, usfm.NodeChapter); len(chapters) != 22 {
		t.Fatalf("chapters = %d, want 22", len(chapters))
	}

	ch1 := findChild(book, usfm.NodeChapter)
	wjCount := 0
	walk(ch1.Children, func(n *usfm.Node) {
		if n.Type == usfm.NodeChar && n.Marker == "wj" {
			wjCount++
		}
	})
	if wjCount == 0 {
		t.Error("no \\wj char nodes in Revelation 1")
	}
}

// BenchmarkParsePsalms measures single-book parse latency on the largest
// book — the number that matters for editor responsiveness (the TS parser
// was ~16ms here; see the perf notes in todo.md).
func BenchmarkParsePsalms(b *testing.B) {
	data, err := os.ReadFile(filepath.Join(bsbDir, "PSA.usfm"))
	if err != nil {
		b.Skipf("BSB corpus not available: %v", err)
	}
	content := string(data)
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		parser.Parse(content)
	}
}

// --- 4. Error characterization ---

func TestAllBooksParseWithZeroErrors(t *testing.T) {
	total := 0
	for _, file := range allUsfmFiles(t) {
		result := parser.Parse(loadUsfm(t, file))
		total += len(result.Errors)
		for i, e := range result.Errors {
			if i >= 3 {
				break
			}
			t.Logf("%s: %s", file, e.Message)
		}
	}
	if total != 0 {
		t.Errorf("total errors across corpus = %d, want 0", total)
	}
}
